src/shared/mkfs-util.c

```diff
@@ -185,6 +188,96 @@ static int do_mcopy(const char *node, const char *root) {
         return 0;
 }
 
+static int protofile_print_item(
+                RecurseDirEvent event,
+                const char *path,
+                int dir_fd,
+                int inode_fd,
+                const struct dirent *de,
+                const struct statx *sx,
+                void *userdata) {
+
+        FILE *f = ASSERT_PTR(userdata);
+        _cleanup_free_ char *base = NULL;
+        int r;
+
+        if (event == RECURSE_DIR_LEAVE) {
+                fprintf(f, "$\n");
+                return 0;
+        }
+
+        if (!IN_SET(event, RECURSE_DIR_ENTER, RECURSE_DIR_ENTRY))
+                return RECURSE_DIR_CONTINUE;
+
+        r = path_extract_filename(path, &base);
+        if (r < 0)
+                return log_error_errno(r, "Failed to extract basename from %p: %m", path);
+
+        char type = S_ISDIR(sx->stx_mode)  ? 'd' :
+                    S_ISREG(sx->stx_mode)  ? '-' :
+                    S_ISLNK(sx->stx_mode)  ? 'l' :
+                    S_ISFIFO(sx->stx_mode) ? 'p' :
+                    S_ISBLK(sx->stx_mode)  ? 'b' :
+                    S_ISCHR(sx->stx_mode)  ? 'c' : 0;
+        if (type == 0)
+                return RECURSE_DIR_CONTINUE;
+
+        fprintf(f, "%s %c%c%c%03o 0 0 ",
+                base,
+                type,
+                sx->stx_mode & S_ISUID ? 'u' : '-',
+                sx->stx_mode & S_ISGID ? 'g' : '-',
```

Comment:
weird way to format that?

Reply:
oh, this was more a comment on xfs chosen syntax for that. not a comment on your code...